mounter/mount.go

```diff
@@ -0,0 +1,377 @@
+	defaultMountCommand = "mount"
+	// FileTypeDirectory defines a constant for the directory FileType
+	FileTypeDirectory FileType = "Directory"
+	// FileTypeFile defines a constant for the file FileType
+	FileTypeFile FileType = "File"
+	// FileTypeSocket defines a constant for the socket FileType
+	FileTypeSocket FileType = "Socket"
+	// FileTypeCharDev defines a constant for the character device FileType
+	FileTypeCharDev FileType = "CharDevice"
+	// FileTypeBlockDev defines a constant for the block device FileType
+	FileTypeBlockDev FileType = "BlockDevice"
+)
+
+// Interface defines the interfact that all mounters must implement
+type Interface interface {
+	// Mount mounts source to target as fstype with given options.
+	Mount(source string, target string, fstype string, options []string) error
+	// Unmount unmounts given target.
+	Unmount(target string) error
+	// List returns a list of all mounted filesystems.  This can be large.
+	// On some platforms, reading mounts is not guaranteed consistent (i.e.
+	// it could change between chunked reads). This is guaranteed to be
+	// consistent.
+	List() ([]MountPoint, error)
+	// IsMountPointMatch determines if the mountpoint matches the dir
+	IsMountPointMatch(mp MountPoint, dir string) bool
+	// IsLikelyNotMountPoint uses heuristics to determine if a directory
+	// is a mountpoint.
+	// It should return ErrNotExist when the directory does not exist.
+	// IsLikelyNotMountPoint does NOT properly detect all mountpoint types
+	// most notably linux bind mounts and symbolic link.
+	IsLikelyNotMountPoint(file string) (bool, error)
+	// GetMountRefs finds all mount references to the path, returns a
+	// list of paths. Path could be a mountpoint path, device or a normal
+	// directory (for bind mount).
+	GetMountRefs(pathname string) ([]string, error)
+}
+
+// HostUtils defines the interface that all HostUtils implementaions must satisfy
+type HostUtils interface {
+	// DeviceOpened determines if the device is in use elsewhere
+	// on the system, i.e. still mounted.
+	DeviceOpened(pathname string) (bool, error)
+	// PathIsDevice determines if a path is a device.
+	PathIsDevice(pathname string) (bool, error)
+	// GetDeviceNameFromMount finds the device name by checking the mount path
+	// to get the global mount path within its plugin directory
+	GetDeviceNameFromMount(mounter Interface, mountPath, pluginMountDir string) (string, error)
+	// MakeRShared checks that given path is on a mount with 'rshared' mount
+	// propagation. If not, it bind-mounts the path as rshared.
+	MakeRShared(path string) error
+	// GetFileType checks for file/directory/socket/block/character devices.
+	// Will operate in the host mount namespace if kubelet is running in a container
+	GetFileType(pathname string) (FileType, error)
+	// MakeFile creates an empty file.
+	// Will operate in the host mount namespace if kubelet is running in a container
+	MakeFile(pathname string) error
```

Comment:
@msau42 I took a quick glance of whether `MakeFile` and `MakeDir` are needed in this interface at all... They are only used by one volume driver -- host path. But I think they are needed because of nsenter -- they both work differently in the nsenter case...  I'll just try to address the open questions with comments.

Reply:
Since it's only used by one plugin, I think it is fine to leave it out of the interface.

Reply:
One annoying thing is that the answer to "what if the path already exists" is different for `nsenter` vs everything else... But since `nsenter` is internal to k/k only, I'm not going to document that behavior. It won't apply to `k/utils/mount`.

Reply:
> Since it's only used by one plugin, I think it is fine to leave it out of the interface.

But I think we have to have it, because the implementation is different for `nsenter`. The host path driver has no notion of when the kubelet is containerized or not, that's abstracted into what version of `HostUtils` was created. So until that is gone, I think we are forced to keep this.

I've opened https://github.com/kubernetes/kubernetes/pull/81739 to capture all of this, though.